### PR TITLE
Load users gists (Phase 1 - public gists only)

### PR DESCRIPTION
**Pull request by shaybeau731**
Currently, a user can click a button and input their GitHub username into a prompt. Their public gists containing .ipynb files will then be listed. Each gist has a button that, when clicked, will open the notebook in a new window. 

I am still trying to think of a way to authorize to GitHub without having to redirect, so this isn't merge ready yet. But any feedback is welcome!

### CODE REVIEW

**Comment by shaybeau731** [comment]
I found a way to load a user's (public and secret) gists without redirecting away from the notebook page. It involves opening a new window for a couple seconds though, which isn't ideal. 
So right now, there is a button to load public gists based on a given username, and another button to load all gists from the authorized user.

**Comment by mreid-moz** [comment]
Nice! I'll review what's here so far and try it out.

File: jupyter-notebook-gist/extensions/gist.js [hunk 1]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+                }
+            });
+
+        })
+    };
+
+    var format_user_gists = function(responseText) { 
+        var body = $('<table/>').addClass("table");
+        var header = $('<tr/>').addClass("row list_header");
+        header.append("<th>"+"Filename"+"</th>");
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: Don't need the `+` for the table headers, you can simply use `header.append("<th>Filename</th>");`

File: jupyter-notebook-gist/extensions/gist.js [hunk 2]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+        header.append("<th>"+"Description"+"</th>");
+        header.append("<th>"+"Last Updated" + "</th>");
+        header.append("<th>"+"Gist URL"+"</th>");
+        body.append(header);
+        var row, button, files, last_updated, pretty_date;
+        var json_response = JSON.parse(responseText);
+        for (var i=0; i<json_response.length; i++) {
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: Style nit: please add spaces around `=` and `<`

File: jupyter-notebook-gist/extensions/gist.js [hunk 3]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+            files = json_response[i].files;
+            // Only load notebook gists 
+            if (!files[Object.keys(files)[0]].filename.endsWith('.ipynb')) continue;
+
+            // Use same date formatting as SaveWidget's _render_checkpoint
+            last_updated = moment(json_response[i].updated_at);
+            var tdelta = Math.ceil(new Date() - last_updated);
+            if (tdelta < utils.time.milliseconds.d){
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: please add a space between the `if` condition and the `{`

File: jupyter-notebook-gist/create_gist.py [hunk 4]
```diff
@@ -240,6 +240,24 @@ def post(self):
...
+    def get(self):
+
+        # Extract access code
+        access_code = self.extract_code_from_args(self.request.arguments)
+
+        # Request access token from github
+        access_token = self.request_access_token(access_code)
+
+        github_headers = { "Accept" : "application/json",
+                            "Authorization" : "token " + access_token }
```
Comment (mreid-moz) [inline jupyter-notebook-gist/create_gist.py]: Please align the quotes with the previous line

File: jupyter-notebook-gist/extensions/gist.js [hunk 5]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+    var format_user_gists = function(responseText) { 
+        var body = $('<table/>').addClass("table");
+        var header = $('<tr/>').addClass("row list_header");
+        header.append("<th>"+"Filename"+"</th>");
+        header.append("<th>"+"Description"+"</th>");
+        header.append("<th>"+"Last Updated" + "</th>");
+        header.append("<th>"+"Gist URL"+"</th>");
+        body.append(header);
+        var row, button, files, last_updated, pretty_date;
+        var json_response = JSON.parse(responseText);
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: If `json_response.length` is zero, we should add a row with colspan=4 saying "no public notebooks found" or similar.

File: jupyter-notebook-gist/extensions/gist.js [hunk 6]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+        })
+    };
+
+    var format_user_gists = function(responseText) { 
+        var body = $('<table/>').addClass("table");
+        var header = $('<tr/>').addClass("row list_header");
+        header.append("<th>"+"Filename"+"</th>");
+        header.append("<th>"+"Description"+"</th>");
+        header.append("<th>"+"Last Updated" + "</th>");
+        header.append("<th>"+"Gist URL"+"</th>");
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: This column loads the "Open" buttons, not the Gist URL. Maybe call it "Load" or "Open" instead?

File: jupyter-notebook-gist/extensions/gist.js [hunk 7]
```diff
@@ -132,11 +141,117 @@ define(function () {
...
+
+            // Create row containing gist information
+            row = $('<tr/>').addClass("list_item row");
+            row.append("<td>" + files[Object.keys(files)[0]].filename + "</td>");
+            row.append("<td>" + json_response[i].description + "</td>");
+            row.append("<td>" + pretty_date + "</td>");
+            // Create button to load notebook 
+            button = $('<button>Load Gist</button>').addClass("btn btn-default btn-sm");
+            button.click({url: json_response[i].html_url}, load_gist_from_click);
+            button.appendTo(row);
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: I think the "Load" button should be inside a `<td>` too.

**Comment by mreid-moz** [comment]
Thinking about it a bit more, opening up a new window for the Github interaction may be the right thing to do in general - if the user has not yet authorized the jupyter-notebook-gist clientid, they'll have to go through that step on the Github page before it will redirect - it might be awkward to squeeze that into the prompt.

That said, when I try to use the "load all user gists" button, it opens a new tab with a URL of `http://localhost:8888/load_user_gists?code=<the code>`, but which displays a blank page.

**Comment by shaybeau731** [comment]
A new window now pops up for the Github authentication. If the user hasn't authenticated yet (or if they're not logged into Github), they'll be prompted to, and once they're done the window should automatically close. If they've already authenticated, the window will still pop up but won't display anything, and it'll close after a couple of seconds.

**Comment by mreid-moz** [comment]
The code looks good.

When I try running it locally, I get an empty popup window that doesn't auto-close. Have you seen that? See screenshot:
screen shot 2016-02-08 at 11 08 49 am

**Comment by shaybeau731** [comment]
Is the modal not loading as well? If it isn't, then it looks like there's something going on with the message event. Either the LoadGistHandler isn't posting the message containing the gists, or the main window isn't receiving the message. I'll look into how to troubleshoot message events.

**Comment by mreid-moz** [comment]
No, I don't see a modal dialog in the notebook either.  The code path that lets you enter a github username and lists their gists works as expected though.

**Comment by shaybeau731** [comment]
Okay, I altered it so that the targetOrigin argument for the call to postMessage is the URL of the main window. Is the modal loading now? If not, could you try calling postMessage in the LoadGistHandler using '*' as the second argument? If that doesn't work either, I'll have to do more digging.

File: jupyter-notebook-gist/create_gist.py [hunk 8]
```diff
@@ -240,6 +240,24 @@ def post(self):
...
+
+        # Request access token from github
+        access_token = self.request_access_token(access_code)
+
+        github_headers = { "Accept" : "application/json",
+                           "Authorization" : "token " + access_token } 
+        
+        response = requests.get("https://api.github.com/gists", headers=github_headers) 
+
+        self.finish("<script>var gists = '"+response.text+"';window.opener.postMessage(gists, window.opener.location);</script>")
```
Comment (mreid-moz) [inline jupyter-notebook-gist/create_gist.py]: This is subject to quoting problems - the issue I was having was due to there being a Gist in my list containing an apostrophe. In other places we've encoded strings with base64, you could do the same here:
`self.finish("<script>var gists = '" + base64.standard_b64encode(response.text) + "';window.opener.postMessage(gists, window.opener.location);</script>")`

File: jupyter-notebook-gist/extensions/gist.js [hunk 9]
```diff
@@ -132,11 +140,133 @@ define(function () {
...
+
+    // Listen for the response message containing the user's gists 
+    // that is sent from the LoadGistHandler
+    var auth_window;
+    window.addEventListener('message', function(event) {
+        if (event.origin != get_base_path()) return;
+        auth_window.close();
+        Jupyter.dialog.modal({
+            title: "All User Gists",
+            body: format_user_gists(event.data),
```
Comment (mreid-moz) [inline jupyter-notebook-gist/extensions/gist.js]: If you change to base64-encoded data, you'll need to change it here as well:
`format_user_gists(window.atob(event.data))`

**Comment by mreid-moz** [comment]
I found the culprit - I had existing gists with `'` characters in their names which broke the string quoting in the response from `LoadGistHandler`.  Sorry, should have caught that during earlier reviews.

In any case, with the changes I mentioned, it all works, and displays private Gists.

Please merge master into this branch, resolve conflicts, and make sure we're still PEP8 compliant, then this will be ready to land!

**Comment by mreid-moz** [comment]
Oh, please also increment the version in `setup.py` since this adds new functionality.

**Pull request merged**